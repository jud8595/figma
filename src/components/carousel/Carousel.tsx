import React, { Children, useState } from 'react';
import styled from 'styled-components';
import { breakpoints } from '../../utility/breakpoint';


export type CarouselProps = {
  children?: JSX.Element | JSX.Element[]
}

export const Carousel: React.FC<CarouselProps> = ({children}) => {

  const [position, setPosition] = useState(0)

  return  (
    <Container role="slider" aria-labelledby="carousel" aria-valuemin={1} aria-valuemax={Children.count(children)} aria-valuenow={position+1}>
      <Cards>
        {
          React.Children.map(children, (child, index) => <Card className={position===index?'active':''}>{child}</Card>)
        }
      </Cards>
    
      <Pagination>
        {
          React.Children.map(children, (_, index) => <Bullet className={position===index?'active':''} onClick={() => setPosition(index)} />)
        }
      </Pagination>
    </Container>
  )
}

const Container = styled('div')`
  margin-top: 64px;
`;

const Cards = styled('div')`
  display: flex;
  justify-content: center;
  gap: 24px;
  margin: 24px;
`;

const Card = styled('div')`
  display: flex;
  :not(.active) {
    display: none;
  }
  
  @media (min-width: ${breakpoints.TABLET}) {
    :not(.active) {
      display: flex;
    }
  }

  @media (min-width: ${breakpoints.LAPTOP}) {
    display: inline-block;
    width: 320px;
  }
`;

const Pagination = styled('ul')`
  text-align: center;
  & > li {
    display: inline-block;
  }

  @media (min-width: ${breakpoints.TABLET}) {
    display: none;
  }
`;

const Bullet = styled('li')`
  height: 8px;
  width: 8px;
  border-radius: 50%;
  margin: 0 10px;
  background-color: ${props => props.theme.colors.neutral3};

  &.active {
    background-color: ${props => props.theme.colors.primary1};
  }
`;

