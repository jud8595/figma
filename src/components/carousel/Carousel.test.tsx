import React from 'react';
import { act, render, screen } from '@testing-library/react';
import Theme from '../Theme';
import { Carousel } from './Carousel';

test('renders a ServiceCard', () => {
  const { container } = render(
    <Theme>
      <Carousel>
        <div>card1</div>
        <div>card2</div>
        <div>card3</div>
      </Carousel>
    </Theme>
  )
  expect(container).toMatchSnapshot()
});

test('jumps to card3 when pagination item n째3 is clicked', async () => {
  render(
    <Theme>
      <Carousel>
        <div data-testid="card1">card n째1</div>
        <div data-testid="card2">card n째2</div>
        <div data-testid="card3">card n째3</div>
      </Carousel>
    </Theme>
  )

  const items = screen.getAllByRole("listitem")
  expect(items[0].className).toContain('active')
  expect(items[1].className).not.toContain('active')
  expect(items[2].className).not.toContain('active')

  expect(screen.getByTestId('card1').parentElement?.className).toContain('active')
  expect(screen.getByTestId('card2').parentElement?.className).not.toContain('active')
  expect(screen.getByTestId('card3').parentElement?.className).not.toContain('active')

  act(() => {
    items[2].click()
  });

  expect(items[0].className).not.toContain('active')
  expect(items[1].className).not.toContain('active')
  expect(items[2].className).toContain('active')

  expect(screen.getByTestId('card1').parentElement?.className).not.toContain('active')
  expect(screen.getByTestId('card2').parentElement?.className).not.toContain('active')
  expect(screen.getByTestId('card3').parentElement?.className).toContain('active')

});
