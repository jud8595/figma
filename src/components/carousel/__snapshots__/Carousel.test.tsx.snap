// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a ServiceCard 1`] = `
<div>
  <div
    aria-labelledby="carousel"
    aria-valuemax="3"
    aria-valuemin="1"
    aria-valuenow="1"
    class="sc-bczRLJ hvPcGN"
    role="slider"
  >
    <div
      class="sc-gsnTZi jQFcaO"
    >
      <div
        class="sc-dkzDqf dijWDS active"
      >
        <div>
          card1
        </div>
      </div>
      <div
        class="sc-dkzDqf dijWDS"
      >
        <div>
          card2
        </div>
      </div>
      <div
        class="sc-dkzDqf dijWDS"
      >
        <div>
          card3
        </div>
      </div>
    </div>
    <ul
      class="sc-hKMtZM gGTovp"
    >
      <li
        class="sc-eCYdqJ fpozIR active"
      />
      <li
        class="sc-eCYdqJ fpozIR"
      />
      <li
        class="sc-eCYdqJ fpozIR"
      />
    </ul>
  </div>
</div>
`;
